### ISSUE TITLE
Strip release executable?

### ISSUE THREAD

**Issue by charlesnicholson** [facebook/buck2#133]
Hi everyone- I'm starting to play around with buck2 and it looks very interesting!

I noticed that the release build (`cargo build -r`) clocks in at 75MB on my M1 MBP. But, `strip` knocks a solid 23MB off of it! I'm guessing that the buck2 executable isn't meant to be dynamically linked _against_, so maybe it's safe to add a release post-build step that runs `strip`?

<img width="497" alt="Screenshot 2023-04-09 at 10 07 29 AM" src="https://user-images.githubusercontent.com/3010295/230777659-cf72dd2f-7d1b-431c-a14c-61b5719edd89.png">

Might be a free win?

Additionally, though, have you done any size auditing? 52MB is still pretty massive for a natively-compiled executable. I wonder if there are more wins lurking in the map file...

**Comment by stepancheg** [comment]
Unstripped binaries produce better stack traces (when panicking or when using normal anyhow errors with `RUST_BACKTRACE=1`). This is not obvious improvement.

**Comment by charlesnicholson** [comment]
Do you expect every user of buck2 to require stack trace functionality? It's not uncommon to ship fully-stripped binaries, and then when investigating issues, attempt to reproduce them on a more instrumented version of the software.

This is a 30% binary size reduction- if you have 1MM buck2 users each downloading it only once, you've saved 1,000,000 * 23MB == 23TB of global internet traffic, and that's not even considering an army of CI robots that will likely download it from somewhere many times per day. If the hope is that 99+% of them won't experience buck2 crashes, that seems like a huge win :)

**Comment by ndmitchell** [comment]
I'd love to think that Buck2 gets to 1MM, but alas, we are probably a way off for now! At the moment we're definitely keen to get proper stack traces, as that ensures that if/when something goes wrong there isn't a need to try reproducing it first.

For our internal builds we use `--strip-debug` and `-S` on Linux/Mac, which strips most debug info (you can't debug into it), but keeps enough for good stack traces. From memory, I think that saved 90% of the space compared to full stripping. That's probably a good point to aim for. CC @themarwhal (see `deploy/buck2_upload.py` for what we do internally).

**Comment by stevefan1999-personal** [comment]
My two cents -- maybe add these to Cargo.toml:

```toml
[profile.release]
incremental = true
lto = "thin"
strip = "debuginfo"

[profile.min-size-release]
inherits = "release"
codegen-units = 1
lto = "fat"
opt-level = "s"

[profile.release-strip]
inherits = "release"
strip = "symbols"

[profile.min-size-release-strip]
inherits = "min-size-release"
strip = "symbols"
```

**Comment by themarwhal** [comment]
Thanks for the suggestions and stats! I really appreciate it. Taking a look at compressing + some baseline stripping today.

**Comment by thoughtpolice** [comment]
The `.text` section of the binary is about `52MB` on my normal release build from earlier (so, 104MB total), the rest going to debug info. Using a tool like [cargo-bloat](https://github.com/RazrFalcon/cargo-bloat) doesn't immediately point anything out to me (actually the biggest symbol is actually a sqlite3 function!); one issue of course is Rust's strong reliance on monomorphization means that a lot of individual symbols will end up being very small, similar copies and so won't stick out like a sore thumb in a "top N" list, it's just death by a thousand cuts. This is one reason why LTO and outlining from `opt-level = z` are relatively effective on any-ol-average Rust (and C++!) codebase in my experience, because those many tiny symbols get deduplicated/outlined/merged/dead-code-eliminated more effectively thanks to global visibility over the code graph. However there's definitely cases where monomorphization is best undone and calls should be left to sit behind a dereference. It's hard to see that sometimes from raw histogram of symbol names, though...

Honestly, text section size is something an afterthought for many things today, not just Rust projects &mdash; specifics of things like monomorphization or larger dependency chains (that have to be more aggressively optimized away with something like LTO) aside. So I'd just guess it was somewhat low on the list. But there are at least a couple easy immediate wins. More aggressive ones would be things like boxing/dyn'ing some dispatches, or even replacing dependencies, but that's probably out of scope for this issue.

**Comment by charlesnicholson** [comment]
Thanks for taking the time to write that response up despite it veering off of the topic for this issue :) I figured the answer was roughly what you described. It's amusing to me that even the fully-stripped undebuggable buck2 executable is the same size as a minimal Debian image! I guess disk space is cheap these days...

**Comment by ndmitchell** [comment]
We have no specific targets for size - stripping and compression seems worthwhile. If you happen to stumble on something that improves code quality and reduces size (and typically reduces compile time - which is always welcome!) then we'd be delighted to take a diff. But we aren't particularly focused on it, and there's no specific target size at which we worry.
